James revisions

jamessimone opened the pull request:
The first commit are small tweaks, some of which may be more stylistic.  The second commit has a few things I wanted to show you, though I take no credit for TestingUtils as that was a joint effort that I have included primarily for clarity in the mock SObjectRepository's methods.

Commit:
Initial commit with small tweaks that hopefully don't break anything

Commit:
Secondary commit with interesting things for Jonathan to see

jamessimone commented:
Oh dear.  Travis seems to be telling me that despite some very careful git snipping, I still managed to break a few things.  I'm sorry!  Hopefully you'll still be able to take a look, with me asking for forgiveness in advance.

jongpie commented:
Hey man - just now getting caught back up on Nebula, had to take a break to deal with some other stuff. When you have time, can you try going a pull from the dev branch into your branch? I think your getting the Travis errors because you're a few commits behind.

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -166,11 +180,12 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
 
     private void addCommonQueryFields() {
         if(!NebulaSettings.repositorySettings.IncludeCommonFields__c) return;
+        if(!this.shouldAddCommonFields) return;
```
I moved the check for 'adding common fields' to a custom setting - do you think it's useful to have both a custom setting and a property/method to control adding the common fields?

jamessimone replied:
the custom setting is almost certainly the better implementation. I haven't had the time to flesh out the required custom settings in my
own org but that's one of my top priorities once I'm back from my trip

Commit:
Merge branch 'dev' into james-revisions

# Conflicts:
#	src/classes/QueryBuilder.cls

Commit:
bringing Nebula up to date with the implemented SObjectRepository and surrounding classes we've been using

Commit:
making QueryOperator static

Commit:
build fixes

Commit:
further build fixes

Commit:
another one!

Commit:
exceptions consolidated

Commit:
forgot to pass QueryFilterScope

Commit:
removing unnecessary overrides in SObjectRepository_Tests and unused exception (has been consolidated into Exceptions.cls)

jamessimone commented:
@jongpie there are a few tests failing in SObjectRepository_Tests that I haven't had the chance to investigate yet, but I have fixed the remainder of the compile issues in this branch.  It is my hope to have the test failures resolved in the near future as well, but I have to run for now - shoot me a line when you've seen this, though!

Edit - very strange.  I fixed the actual issues with the tests and have them all passing on my local, which is a totally new sandbox I just made.  When pushing changes from my branch, travis initially will report that the build has failed but that all of the tests have passed - when I click on the "details" below to look at the most recent builds, though, there are 4 tests that are allegedly failing:

it_should_update_a_single_record
it_should_search_in_all_fields
it_should_return_new_this_week_by_id
it_should_return_my_recent_leads

It'd be great to get a fresh set of eyes on this issue - I'm sure that it's some kind of environmentally induced bug, but I wasn't able to track it down tonight.

jamessimone committed:
test fixes

jamessimone committed:
fixed one other test while trying to figure out why travis build is failing

jamessimone committed:
adding test coverage and fixed a few issues

jamessimone commented:
@jongpie  this is now ready to be reviewed properly

jongpie commented:
Hey man! Sorry for the delay in reviewing this, I'll have it done by this weekend!

jamessimone commented:
don't worry about it dude.  it's a big PR.  I would definitely recommend in particular looking at how the SObjectRepositoryMocks hook into CrudMock so that DML operations can be tested against explicit lists in testing utils.  Again, we have 241 tests and they are currently hovering around 1 minute to all run - I think they run faster when run synchronously, as well.

jamessimone commented on src/classes/QueryParentPredicate.cls:
```diff
@@ -0,0 +1,31 @@
+public class QueryParentPredicate extends QueryPredicate {
```
this is the addition for querying parent fields.  Because our implementation of QueryBuilder and SObjectRepository are already relying heavily on a generic "get" method by List<QueryPredicate>, pushing the implementation to the predicate felt right.  

The utility of the predicate really has me feeling like I would actually prefer to rip out the other get methods in the interface and rely fully on the predicate.

jongpie replied:
@jamessimone yo dawg, can you back out these changes for now and put a hold on your PR? I'm getting lost trying to keep up with your revisions and I've already got a solution in the works for the parent field filtering

jamessimone commented on src/classes/SObjectRepository.cls:
```diff
@@ -2,40 +2,98 @@
     private SObjectRepository(Schema.SObjectType sobjectType, Schema.FieldSet fieldSet, List<Schema.SObjectField> sobjectFieldList) {
         this.currentModule = NebulaCore.Module.REPOSITORY;
 
-        this.queryBuilder = new QueryBuilder(sobjectType, fieldSet, sobjectFieldList);
```
removed in favor of line 30, below - fixed a nasty bug where the query builder was saving state if multiple calls to the repository was made in the same transaction

jongpie replied:
Oooh, nice catch & fix :+1:

jamessimone commented on src/classes/SObjectRepository.cls:
```diff
@@ -2,40 +2,98 @@
+            .getQueryResults();
+    }
+
+    public virtual List<SObject> getByFieldForIds(Schema.SObjectField field, Object value, Set<Id> idSet) {
```
lines 59 - 83 are the ones I am thinking seriously about replacing with query predicates in the calling implementations.

jongpie commented on src/classes/Crud.cls:
```diff
@@ -0,0 +1,53 @@
+* This file is part of the Nebula Framework project, released under the MIT License.             *
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public virtual class Crud implements ICrud {
```
I like this class, but I think it should be called DML instead of CRUD (both to match other apex commands and because some DML commands aren't standard CRUD operations)

jongpie commented on src/classes/CrudMock.cls:
```diff
@@ -0,0 +1,42 @@
+* This file is part of the Nebula Framework project, released under the MIT License.             *
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public virtual class CrudMock implements ICrud {
```
Same here, I'd vote for using 'DML' over 'CRUD' in the name

jongpie commented on src/classes/DescribeUtils.cls:
```diff
@@ -0,0 +1,28 @@
+public without sharing class DescribeUtils {
+
+    public static SObjectType getSObjectType(Schema.SObjectField field) {
+        //There is supposed to be a built in Salesforce getDescribe() result for Schema.SObjectField, getReferenceTo()
```
Haha, I found the same thing on Stack Exchange. I started working on a similar class, I think I've got a few extra features in my version - you mind removing this class for now? I'll add mine to the repo after we rap up this PR.

jongpie commented on src/classes/CollectionUtils.cls:
```diff
@@ -0,0 +1,15 @@
+        return listOfThings == null || listOfThings.isEmpty();
+    }
+
+    public static SObject getFirstOrDefault(List<SObject> listOfThings) {
```
I don't like this method - returning null doesn't seem like the same thing as a returning some kind of default. I vote for scrapping this guy.

jongpie commented on src/classes/IQueryBuilder.cls:
```diff
@@ -5,19 +5,22 @@
 
     // Query execution methods
-    SObject getFirstQueryResult();
+    SObject getFirstOrDefault();
```
See comment on the CollectionUtils class, but I don't like the idea of returning a default (especially since right now, the code just returns null). There's no concept in Salesforce or Apex of a default record, so I don't think it makes sense to introduce it into the framework.

jongpie commented on src/classes/IQueryBuilder.cls:
```diff
@@ -5,19 +5,22 @@
     List<SObject> getQueryResults();
     List<SObject> getSearchResults(String searchTerm, QuerySearchGroup searchGroup);
-
+    List<AggregateResult> getAggregate(List<String> selectAliases, List<SObjectField> groupByFields,
```
Let's back out the aggregate stuff for now, dawg, and sort that out later - you're killing me with having 1 massive PR for all of your changes :-(

jamessimone replied:
sorry man.  I should have just forked the repo.  Mostly this was just so you could see what we have been doing with it - I definitely agree, the branch has gotten out of hand

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -2,41 +2,54 @@
-        this.whereClauseList     = new List<String>();
-        this.orderByList         = new List<String>();
-        this.forUpdate           = false;
+        this.shouldAddCommonFields = true;
```
Didn't we talk about this a few weeks ago? There's a custom setting that handles adding the common fields - we should remove this variable & the related methods.

jamessimone replied:
sorry.  some of the changes you are seeing (which I will roll back) are because:
a) we haven't implemented the Logger.cls yet
b) we haven't implemented guard clauses for custom settings.  

That's totally my bad and something I meant to mention earlier.

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -70,107 +83,179 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
-    public SObject getFirstQueryResult() {
-        return this.getQueryResults()[0];
+    // Query execution methods
+    public SObject getFirstOrDefault() {
```
I made a few other comments about this elsewhere, but I vote to undo the name change here

jongpie commented on src/classes/SObjectRepository.cls:
```diff
@@ -2,40 +2,102 @@
+            .getQueryResults();
+    }
+
+    public List<SObject> getByFieldAndTypeForGivenTimePeriod(Schema.SObjectField field, Object firstValue, Object secondValue) {
```
I think the method name & variable names are confusing.
- getByFieldAndTypeForGivenTimePeriod - just given the name, it seems like a pretty specific use case anyway. You sure it belongs in the framework? If so, that's cool, just want to be sure.
- 'firstValue' - maybe 'fieldValue' instead?
- 'secondValue' - 'createdDate'? 'createdAfter'? 'createdSince'?

jamessimone replied:
no, I will remove this.

jongpie commented on src/classes/QueryPredicate.cls:
```diff
@@ -0,0 +1,60 @@
+* This file is part of the Nebula Framework project, released under the MIT License.             *
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public class QueryPredicate {
```
I'm struggling on this guy a bit - what benefit does it add? I don't see unit tests or example implementations to go off of, but it seems like using this approach would add more overhead than just calling the querybuilder's whereField method.

jamessimone replied:
I think the benefit is that with the query builder's methods not (previously) being accessible (since they have been hidden behind the SObjectrRepository), we previously were making many highly specific methods in the repository class / interface related to getting things.  with the query predicates, you can encapsulate a whole list of criteria within one single "get" request.

Here's an example:
`QueryPredicate emailInAddresses = QueryPredicate.EQUALS(Schema.Lead.Email, contactEmails);
QueryPredicate notClosed = QueryPredicate.EQUALS(Schema.Lead.IsConverted, false);`
List<SObject> leads = this.leadRepo.get(new List<QueryPredicate> {emailInAddresses, notClosed});

jongpie replied:
Ok, that makes total sense and seems like a pretty elegant way to handle. But now I'm curious about your implementation - do you still create methods in LeadRepo for your generic/commonly used queries? Or are you just using "get" request (QueryPredicate) for everything?

jamessimone replied:
we have switched to using get for pretty much everything because we are often constructing objects with several repos and passing their interfaces in accordingly, which obviates the need for having specific types in the objects being constructed themselves - but that would be pretty hard to manage if we had to keep on adding methods to the base interface. the query predicate encapsulates the request nicely, and the repo is just responsible for query / dml operations. ALSO (bonus): you can pass an empty query predicate to get everything (which we have a use case for: querying a very small object that serves as a joint object
between User and sales people that serves as the master in a master detail relationship)

jongpie replied:
As the Spanish always say (probably), "muy sexy." 

I'm sold, seems like a super nice approach. Imma re-re-review the rest of the PR now that you've convinced me on this part.

jamessimone replied:
Cool!  Let me know what else you think.  Now that we have query predicates, any time we create a new object, it's easy for us to spin up a new repo with the virtual SObjectRepository, a short list of fields, and our factory class that handles actual object creation.  We did that today with a new joint object and it took all of a minute.  Even if sometimes we're using the repo just for the crud / dml operations, you can definitely see the progression as we move towards defining the boundaries - and objects that were previously just being inserted begin to take on new logic as we find the need to check inserted / updated values.  Honestly, between the trigger framework and the SObjectRepository (to say nothing of the mocks, which have sped up testing time to mere seconds), it's been fun to imagine and then easily implement the business logic we're building out.  

Looking forward to seeing your take on the querying parent objects, as well!

jongpie replied:
It sounds pretty magical that you implemented a new object's classes in about a minute :+1:

jongpie commented on src/classes/TestingUtils.cls:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    public static SObject setReadOnlyField(SObject sobj, String fieldName, String value) {
```
I love these methods, but thoughts on using Schema.SObjectField instead of String for field names? I've been trying to avoid relying on strings in the framework as much as possible (typos in strings have been one of many of my nightmares lately - SObjectField would prevent typos in field names when writing tests).

jamessimone replied:
absolutely - definitely a great call.  Type safety is very important!

jongpie commented on src/classes/TestingUtils.cls:
```diff
@@ -0,0 +1,40 @@
+* This file is part of the Nebula Framework project, released under the MIT License.             *
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public class TestingUtils {
```
This class is pretty baller. Since it's meant to be used only in tests, what do you think about adding something like this to it?
```
public class TestingUtils {
    static {
        if(!Testing.isRunningTest()) throw new Exceptions.InvalidOperationException('Nah, no way, sorry');
    }
...
```
This would prevent someone from using anything in the TestingUtils class - it would only run in a test context. What do you think - overkill?

jamessimone replied:
no that definitely makes sense!

jongpie commented on src/classes/TestingUtils.cls:
```diff
@@ -0,0 +1,40 @@
+            serializedRecord += ',"' + stringer + '":"' + changesToFields.get(stringer)+'"';
+        }
+
+        serializedRecord+= + '}';
```
The plus signs in this line hurt my brain - I get what it's doing, but it's a little disorienting to read at first. Anyway to tweak it a bit to be more readable?

jamessimone replied:
I tried to make a little more readable by inserting line breaks, but it's pretty ugly either way I think :-\

jongpie commented on src/classes/SObjectRepositoryMocks.cls:
```diff
@@ -0,0 +1,114 @@
+        public virtual List<SObject> getById(List<Id> recordIdList) {
+            this.recordIdList = recordIdList;
+
+            if(this.records == null) {
```
Can we smoosh this down to one line?
`if(this.records == null) this.createMockObjects();`

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -2,41 +2,54 @@
 * See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
 *************************************************************************************************/
-public without sharing class QueryBuilder extends NebulaCore implements IQueryBuilder {
+public without sharing virtual class QueryBuilder extends NebulaCore implements IQueryBuilder {
```
Do you have specific plans for extending the class in your implementation? Or do you think we should make all the main classes in the framework be virtual?

jamessimone replied:
at one point we had a test class extending QueryBuilder to inject a mock, but that has since been deprecated!  I will back out that change - looks like we need to do the same!

jamessimone replied:
ah yes, I remember now - I had an extension for returning multiple objects with SOSL in the first implementation of our "make email address unique" story.  We ended up going with two SOQL calls from Lead / Contact repo's instead.  Apologies again, backing this out

jongpie commented on src/classes/QueryDateLiteral.cls:
```diff
@@ -4,185 +4,194 @@
     //Actual constant literals
-    public QueryDateLiteral YESTERDAY {
-        get {return this.setValue('YESTERDAY');}
+    public static QueryDateLiteral YESTERDAY {
```
Switching to static definitely makes more sense here - I knew something was bugging me about this class still :+1:

jamessimone replied:
haha I just realized QueryArgumentFormatter still needs to be converted, but that can be done at a later time

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -70,107 +83,179 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
     }
 
     private void addCommonQueryFields() {
-        if(!NebulaSettings.repositorySettings.IncludeCommonFields__c) return;
```
Undo this change - let's stick with the custom setting

jongpie commented on src/classes/SObjectRepository.cls:
```diff
@@ -2,40 +2,102 @@
+    //SOQL
+    public virtual SObject getById(Id recordId) {
+        return this.queryBuilder
+            .whereField(new List<QueryPredicate>{QueryPredicate.EQUALS(this.idField, recordId)})
```
Related to my confusion on the benefits of QueryPredicate, line 42 seems more complex than the original way. This seems easier to ready to me:
`whereField(this.idField, QueryOperator.EQUALS, recordId)`

jamessimone replied:
Your version is definitely easier to read.  And we had it like that for a while.  Again, I think the advantage of using the query predicates is that you don't have to constantly be peering into the QueryBuilder to see what's going on / what's supported - you can just pass in a query predicate or two and know that whatever you toss its way is going to get handled.  Until we hit the stopper of trying to query parent fields, I hadn't looked at the implementation in quite a while. 

I'm reverting this!

jongpie commented on src/classes/SObjectRepository.cls:
```diff
@@ -2,40 +2,102 @@
+    }
+
+    public List<SObject> getCreatedSinceTimeValue(Object timeValue) {
+        Schema.SObjectField createdDate = this.getCreatedDate();
```
I'd scrap this line (and the variable) and just call the method directly in the whereField call
`.whereField(this.getCreatedDate(), QueryOperator.GREATER_THAN_OR_EQUAL_TO, timeValue)`

jongpie commented on src/classes/SObjectRepositoryMocks.cls:
```diff
@@ -0,0 +1,114 @@
+            //Setup record and fields to be returned
+            SObjectType sobjType = DescribeUtils.getSObjectType(field);
+            SObject record = sobjType.newSObject();
+            record.Id = TestingUtils.generateId(record.getSObjectType());
```
This is some good stuff :+1:

jongpie commented on src/classes/QueryAggregateCommand.cls:
```diff
@@ -0,0 +1,8 @@
+public enum QueryAggregateCommand {
```
Same here - let's hold off on tackling new features like aggregates until we get your revisions wrapped up.

jamessimone replied:
yup for sure.  I will back this stuff out, hopefully tomorrow

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -70,107 +83,179 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
         return this;
     }
 
+    public IQueryBuilder turnOffCommonFields() {
```
Remove/let's go back to the custom setting

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -70,107 +83,179 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
+        this.query += String.join(queryStringSpecifiers, ',');
+    }
+
+    private String getFieldNameByRelationship(SObjectField field) {
```
Mind removing this? I've got some other code that in progress that handles this a bit differently.

jamessimone replied:
not a problem at all - can't wait to see what you're building

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -70,107 +83,179 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
-        if(this.filterScope != null) Logger.addEntry(this, 'SOSL Search Query method flagged as USING SCOPE ' + this.filterScope + '. SOSL cannot use USING SCOPE, ignoring');
-
-        Logger.addEntry(this, query);
+        if(this.forUpdate) System.debug(LoggingLevel.WARN, 'SOSL Search Query method flagged as FOR UPDATE. SOSL cannot use FOR UPDATE, ignoring');
```
I'd prefer to use the Logger class instead - what was the thought behind switching to System.debug instead?

jamessimone replied:
Ah again - this is just my bad.  I missed that once in a selective staging many commits ago and haven't backed it out.  Because we're not using the whole framework (mostly just the SObjectRepo and associated classes) there are some inconsistencies in the files - inconsistencies that have affected you in readability, so I am very sorry about that!

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -182,19 +267,26 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
         if(this.sobjectFieldList == null) return;
 
-        for(Schema.SObjectField field : this.sobjectFieldList) this.queryFields.add(field.getDescribe().getName().toLowerCase());
+        for(Schema.SObjectField field : this.sobjectFieldList) {
```
Can we smoosh this back down to 1 line?
`for(Schema.SObjectField field : this.sobjectFieldList) this.queryFields.add(field.getDescribe().getName().toLowerCase());`

jongpie commented on src/classes/QueryBuilder.cls:
```diff
@@ -182,19 +267,26 @@ public without sharing class QueryBuilder extends NebulaCore implements IQueryBu
         if(this.fieldSet == null) return;
 
-        for(Schema.FieldSetMember field : this.fieldSet.getFields()) this.queryFields.add(field.getFieldPath().toLowerCase());
+        for(Schema.FieldSetMember field : this.fieldSet.getFields()) {
```
Can we smoosh this back down to 1 line?
`for(Schema.FieldSetMember field : this.fieldSet.getFields()) this.queryFields.add(field.getFieldPath().toLowerCase());`

jongpie reviewed (changes requested):
Hey man - I finally had some time to review the code this weekend. Overall, it looks pretty nice, but I've added several comments throughout the changes. The biggest things are:
1. Can you roll back your changes related to aggregate results & parent fields? This PR has gotten out of hand and it's hard to review your tweaks when you're also adding big new features into the mix (and I've already got a solution going for the parent fields).
2. QueryPredicate class confuses me - I'm sure I'm missing something, but I'm struggling with understanding the use case.

I think for now, most everything else is related to smaller questions, etc.

jamessimone committed:
code review changes

jongpie commented on src/classes/QueryPredicate.cls:
```diff
@@ -0,0 +1,60 @@
+* This file is part of the Nebula Framework project, released under the MIT License.             *
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public class QueryPredicate {
```
Ok, that makes total sense and seems like a pretty elegant way to handle. But now I'm curious about your implementation - do you still create methods in LeadRepo for your generic/commonly used queries? Or are you just using "get" request (QueryPredicate) for everything?

jongpie commented on src/classes/DML.cls:
```diff
@@ -0,0 +1,53 @@
+    public virtual void doHardDelete(SObject record) { this.doHardDelete(new List<SObject>{record}); }
+    public virtual void doHardDelete(List<SObject> records) {
+        this.doDelete(records);
+        if(records.isEmpty() == false) {
```
Small thing, but it looks like you flip between these 2 options in some of your code

1. `someList.isEmpty() == false`
2. `!someList.isEmpty()` (line 26 & 39 in DMLMock_Tests.cls and a few other places)

I prefer option 2 myself, but I could do either. I just want to try to be consistent as much as possible & try to stick with just one of them. Thoughts? Preferences?

jamessimone replied:
I prefer option 2, as well - but my team prefers option 1.  So there is some mixed styling where I made some changes with them.  I will fix those.

jamessimone committed:
fixing awkward .isEmpty() checks

jongpie commented:
Ok, I've been reviewing your latest changes and I think things are looking awesome (you've won me over on QueryPredicate, it's awesome).

2 last changes though (I promise this is it!):
1. Do you miind deleting the file  .vim-force.com/session.properties? (And did you stop using MavensMate?)
2. Travis says it's failing because code coverage is at 74% / any chance of adding enough to cover that last percent?

jamessimone committed:
fixing a slight bug that came up while testing predicates with schedulable interfaces today - a null reference exception could be thrown in certain instances

jongpie commented on src/classes/QueryBuilder_Tests.cls:
```diff
@@ -0,0 +1,58 @@
+        System.assert(queryFactory.orderByList[0].contains(Contact.CreatedDate.getDescribe().getName()));
+    }
+
+    @isTest
```
* This method is empty, I suggest to remove it.
* Test methods must use assertions to ensure logic works as expected.

jongpie replied:
* Test methods must use assertions to ensure logic works as expected.
* This method is empty, I suggest to remove it.

jongpie commented on src/classes/DML.cls:
```diff
@@ -0,0 +1,53 @@
+* See LICENSE file or go to https://github.com/jongpie/NebulaFramework for full license details. *
+*************************************************************************************************/
+public virtual class DML implements IDML {
+    public virtual SObject doInsert(SObject record) { return this.doInsert(new List<SObject>{record})[0];}
```
@jamessimone sorry, another question here! I swear we discussed this a month or two ago, but why return SObject (or a list of them) on these methods? The variables passed to the methods should already get updated to have the IDs after the insert occurs, so I don't think there's a need to return anything (plus then you have to cast them from SObject back to the appropriate concrete SObject type).

jamessimone replied:
@jongpie  this was a change suggested by Marty to come in line with the return types for the actual DML statements themselves.  I agree, typically you wouldn't actually be re-updating your variables to be equal to what is being returned.  More of being consistent with the larger salesforce API

jongpie commented on src/classes/QueryPredicate.cls:
```diff
@@ -0,0 +1,60 @@
+    private final QueryOperator specificOperator;
+    private final List<Object> specificValues;
+
+    public static QueryPredicate EQUALS(Schema.SObjectField field, Object value) {
```
@jamessimone sorry again, but more questions now that I've had a bit more sleep. Is there any difference between..
`QueryPredicate.EQUALS(Lead.Id, expectedLead.Id)`
and this?
`new QueryPredicate.(Lead.Id, QueryOperator.EQUALS, expectedLead.Id)`

It seems like they do the same thing, right? (Although the second option isn't technically possible right now because the constructors are private)

jamessimone replied:
@jongpie the reason the constructors are private is because you can pass EITHER QueryPredicate.Equals(Lead.Id, expectedLead.Id) OR QueryPredicate.EQUALS(Lead.Id, new List<Id>{expectedLead.Id}) - with the point being that if you were to replace your second variable with ANOTHER list of Id's, the combination of predicate and QueryArgumentFormatter will work together in blissful harmony to decide whether or not to use = 'id here' or: IN (List of comma separated Id's here)

jamessimone committed:
adding test coverage

jamessimone committed:
removing commented out test

jamessimone commented:
@jongpie  the only thing preventing merging at this point is clayton, I believe.  I've taken care of everything you suggested!

jongpie commented:
@jamessimone sounds good! I'm not too worried​ about the complaints from Clayton - do you want me to override it and go ahead and merge this PR?

jongpie commented:
@jamessimone nevermind! You should be able to merge now - I changed Clayton to be an optional check on the repo for now. 

I've also switched the repo to only use squash for merging - I think that eventually, I will delete old feature branches to keep the repo clean and using squash should help with that.

jamessimone commented:
@jongpie just merged - let me know if you are going to work on the readme at all, I would like to write up something about the mocks and how to reduce overall testing time dramatically by using them.

jongpie commented:
@jamessimone woohoo, it's merged! :1st_place_medal: :+1: :100: 

If you want to start on the README, that would actually be awesome - I've been wanting to work on that, but I haven't had a chance to work on it yet.

I also created a wiki within github (let me know if you don't see it, not sure if I need to tweak the permissions). At some point, I want to put more detailed documentation in there (and keep readme as more general info) - not sure how detailed your mock info will be, so feel free to put it wherever you think it makes sense.
